sonar-html-plugin/src/main/java/org/sonar/plugins/html/checks/accessibility/DomElementsShouldUseAutocompleteAttributeCorrectlyCheck.java

```diff
@@ -0,0 +1,323 @@
+/*
+ * SonarSource HTML analyzer :: Sonar Plugin
+ * Copyright (c) 2010-2024 SonarSource SA and Matthijs Galesloot
+ * sonarqube@googlegroups.com
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.sonar.plugins.html.checks.accessibility;
+
+import org.sonar.check.Rule;
+import org.sonar.plugins.html.api.HtmlConstants;
+import org.sonar.plugins.html.api.accessibility.ControlGroup;
+import org.sonar.plugins.html.checks.AbstractPageCheck;
+import org.sonar.plugins.html.node.TagNode;
+
+import java.util.List;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+
+@Rule(key = "S6840")
+public class DomElementsShouldUseAutocompleteAttributeCorrectlyCheck extends AbstractPageCheck {
+  static boolean isADirective(String value) {
+    return value.startsWith("<%=") || value.startsWith("<?");
+  }
+
+  static class Validator {
+    protected List<Predicate<Candidate>> constraints;
+
+    public Validator(List<Predicate<Candidate>> constraints) {
+      this.constraints = constraints;
+    }
+
+    boolean isValid(TagNode tagNode) {
+      String autocomplete = tagNode.getAttribute("autocomplete");
+
+      if (autocomplete == null) {
+        return true;
+      }
+
+      String[] tokens = isADirective(autocomplete) ? new String[]{autocomplete} : autocomplete.split(" ");
+
+      if (tokens.length != this.size()) {
+        return false;
+      }
+
+      boolean isValid = true;
+
+      for (int i = 0; i < constraints.size(); i++) {
+        Predicate<Candidate> constraint = constraints.get(i);
+        String token = tokens[i];
+        Candidate candidate = new Candidate(token, tagNode);
+
+        isValid = isValid && constraint.test(candidate);
+      }
+
+      return isValid;
+    }
+
+    int size() {
+      return constraints.size();
+    }
+  }
+
+  static class Token {
+    private final String value;
+
+    private final Predicate<TagNode> controlGroupPredicate;
+
+    Token(
+      String value,
+      Predicate<TagNode> controlGroup
+    ) {
+      this.value = value;
+      this.controlGroupPredicate = controlGroup;
+    }
+  }
+
+  static class Candidate {
```

Comment:
Let's document what we mean by candidate with a sentence or two and an example.

```suggestion
  /** ... */
  static class Candidate {
```